File: packages/shared-types/python/src/mycelium_shared_types/transitions.py
"""AgentTask lifecycle transitions. The single source of truth."""

from __future__ import annotations

from mycelium_shared_types.agent import AgentTaskStatus

VALID_AGENT_TASK_TRANSITIONS: dict[AgentTaskStatus, set[AgentTaskStatus]] = {
    AgentTaskStatus.QUEUED: {
        AgentTaskStatus.RUNNING,
        AgentTaskStatus.PENDING_APPROVAL,
        AgentTaskStatus.CANCELLED,
    },
    AgentTaskStatus.PENDING_APPROVAL: {
        AgentTaskStatus.RUNNING,
        AgentTaskStatus.FAILED,
        AgentTaskStatus.CANCELLED,
    },
    AgentTaskStatus.RUNNING: {
        AgentTaskStatus.SUCCEEDED,
        AgentTaskStatus.FAILED,
        AgentTaskStatus.PENDING_APPROVAL,
    },
    AgentTaskStatus.FAILED: {AgentTaskStatus.RETRIED, AgentTaskStatus.CANCELLED},
    AgentTaskStatus.RETRIED: {AgentTaskStatus.SUCCEEDED, AgentTaskStatus.CANCELLED},
    AgentTaskStatus.SUCCEEDED: set(),
    AgentTaskStatus.CANCELLED: set(),
}


def is_valid_agent_task_transition(
    current: AgentTaskStatus, next_status: AgentTaskStatus
) -> bool:
    """Return True if ``current → next_status`` is a permitted lifecycle transition."""
    return next_status in VALID_AGENT_TASK_TRANSITIONS.get(current, set())

File: packages/shared-types/typescript/src/transitions.ts
import type { AgentTaskStatus } from "./agent.js";

/** Single source of truth for AgentTask lifecycle transitions. */
export const VALID_AGENT_TASK_TRANSITIONS: Record<AgentTaskStatus, AgentTaskStatus[]> = {
  queued: ["running", "cancelled"],
  running: ["succeeded", "failed"],
  failed: ["retried", "cancelled"],
  retried: ["succeeded", "cancelled"],
  succeeded: [],
  cancelled: [],
};

export function isValidAgentTaskTransition(
  current: AgentTaskStatus,
  next: AgentTaskStatus,
): boolean {
  return VALID_AGENT_TASK_TRANSITIONS[current]?.includes(next) ?? false;
}
